File: leetcode-go/23.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeLists(l1, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	}
	if l1.Val > l2.Val {
		temp := l2
		l2 = l1
		l1 = temp
	}
	temp := l1.Next
	result := l1
	result.Next = nil
	l1 = temp
	traverseNode := result
	for l1 != nil && l2 != nil {
		insertNode := l1
		temp = l1.Next
		if l1.Val > l2.Val {
			temp = l2.Next
			insertNode = l2
			l2 = temp
		} else {
			l1 = temp
		}
		insertNode.Next = nil
		traverseNode.Next = insertNode
		traverseNode = traverseNode.Next
	}
	if l2 == nil && l1 != nil {
		traverseNode.Next = l1
	} else if l1 == nil && l2 != nil {
		traverseNode.Next = l2
	}
	return result
}

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}
	if len(lists) == 1 {
		return lists[0]
	}
	listOfLists := []*ListNode{}

	for len(lists) > 1 {
		index := 0
		for index+1 < len(lists) {
			mergedList := mergeLists(lists[index], lists[index+1])
			// fmt.Println(mergedList, mergedList.Next)
			listOfLists = append(listOfLists, mergedList)
			index += 2
		}
		if index+1 == len(lists) {
			listOfLists = append(listOfLists, lists[len(lists)-1])
		}
		// fmt.Println(len(listOfLists))
		lists = listOfLists
		listOfLists = []*ListNode{}
	}
	return lists[0]
}

File: leetcode-js/23.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

function merge(list1, list2) {
    if (!list1 && !list2) return null;
    if (!list1) return list2;
    if (!list2) return list1;

    if (list1.val >= list2.val) {
        let temp = list1;
        list1 = list2;
        list2 = temp;
    }

    let result = list1;
    let currNode = list1;
    list1 = list1.next;
    while (list1 && list2) {
        if (list1.val <= list2.val) {
            currNode.next = list1;
            list1 = list1.next;
        }
        else {
            currNode.next = list2;
            list2 = list2.next;
        }
        currNode = currNode.next;
    }
    if (!list1) currNode.next = list2;
    else currNode.next = list1;
    return result;
}

var mergeKLists = function (lists) {
    if (lists.length === 0) return null;
    let toMerge = lists;
    let newToMerge = [];
    while (toMerge.length > 1) {
        for (let i = 0; i + 1 < toMerge.length; i += 2) {
            let list1 = toMerge[i];
            let list2 = toMerge[i + 1];
            newToMerge.push(merge(list1, list2));
        }
        console.log(toMerge.length);
        if (toMerge.length % 2 === 1) {
            let list = toMerge[toMerge.length - 1];
            newToMerge.push(list);
        }
        toMerge = newToMerge;
        newToMerge = [];
        // console.log(toMerge, toMerge.length);
    }
    return toMerge[0];
};